doc/developer/design/20210831_correctness.md

```diff
@@ -0,0 +1,124 @@
+Update streams that are definite form the "durable storage" of Materialize, and are the basis for Materialize's durability guarantees.
+
+The **compute layer** (`dataflow`, plus potentially others) consumes streams of updates to collections, and computes and then incrementally maintains the results of views over these data.
+A view is ideally "deterministic" in the sense that it acts as a function on the input update streams, and produces a well-defined output on definite input (we can allow "non-deterministic" views but like "indefinite streams" they taint their results).
+One correctness goal of the `dataflow` module is to convert definite inputs into definite outputs, which can be returned to the storage layer or to interactive user sessions.
+Views and their dependencies form the "concurrency control" of Materialize, in that outputs are reported as definite only when, and ideally as soon as, their inputs are definite.
+
+The **coordination layer** (`coord`, plus potentially others) interfaces with user sessions and pilots the storage and compute layers.
+It maintains mappings from user names to storage and compute objects (source, views, indexes), and dereferences them in ways that provide a consistent experience for users.
+It also chooses timestamps for reads (peeks) and writes (table inserts) to provide the lived experience of continually moving forward through time.
+It pilots the storage and compute layers and is responsible for ensuring they communicate clearly and completely with each other.
+One correctness goal of the `coord` module is to provide "strict sequential consistency" to the interactive users, through the careful use of the storage and compute layers.
+Another correctness goal of the `coord` module is to ensure that the storage and compute correctly and completely interoperate (e.g. that the compute layer is invoked only on definite inputs, and fully communicates its definite outputs back to storage if appropriate).
+
+### Approaches
+
+The plan of record is that source data are made definite by the storage layer, and results are definite when observed through the compute layer (applied to definite inputs).
+Definite data are both durable, and ordered in time (strictly sequentially consistent), and reflect the stated "order" the system presents as having undergone.
+
+The coordination layer has several constraints on its behavior to present as strictly sequentially consistent.
+The coordination layer should only report definite data, or indefinite data that are clearly presented as such.
+The coordination layer's catalog should advance in a similar timeline as the data it references, to avoid apparent mis-ordering of DDL and DML statements.
```

Comment:
Is there an implicit assumption that the coordinator makes the catalog durable? Or does it rely on `persist` to do so? I'm pretty sure what the answer is, but it's not clear from this text.

Reply:
That's a good question! I personally think of the "coordinator" as being in charge durability for its catalogs, and gets no help from `persist` unless it wants to model the catalog as a time-varying collection, but the whole thing seems very potentially debatable. A similar Q is about the durability of the `Timeline` struct, which is what provides the linearizability guarantees (time only goes forward, and the coordinator shouldn't report certain things until that forward step is durable). 

I have some unconventional thoughts about the catalog, and maybe it is possible to shake those out independent of the rest of this. Roughly, that the coordinator is probably two parts: "catalog/sql" + "control plane" where "catalog/sql" is some SQL specific stuff, abstractions, behavior, and "control plane" is less sql specific stuff about the correct execution of the system (which dataflows are running now, which persist names correspond to which dataflows; how to turn things on again if they go down). I don't know how these takes shake out, short/medium/long term.

Reply:
It's not for instance clear to me that there aren't problems that arise when you have two different systems keeping different subsets of things durable. Maybe that's ok, but maybe it's not? The sequences of crash faults might have some subsets of data "durable" but not deterministically replayable, and getting that right seems tricky - compared to having them all use a single in-order-sequence of durable things that mixes both the data and metadata in a single stream. I'm not the expert here, but someone should be.

Reply:
I didn't change this very much / at all in the most recent push. I still have the position of not having a strong position that Materialize must do a certain thing with its catalog to be "correct" (at least in the sense of this document). In my head there is a split between "correctness" (this doc) and "consistency" (another doc that allows various SQL weirdness). If anything I'd be happiest further shading the opinions on SQLism in this doc, rather than trying to nail them down here.